### DISCUSSION TITLE
Testing Library의 도입에 대해서

### DISCUSSION THREAD

**Discussion by bbearcookie** [dnd-side-project/dnd-10th-4-frontend#221] [General]
@easyhyun00 

안녕하세요!
Testing Library의 도입에 대해서 논의하고 싶은 내용이 있어서 글을 올렸어요!

지속 개발하면서 새로운 기능을 추가하다 보면 자연스럽게 코드의 복잡성이 오르고,
기존의 코드를 리팩토링 하다 보면 이전에 작성했던 기능이 정말로 정상 작동하는지 체크하는 과정이 많아지는데요..

수동으로 개발 서버를 띄워서 작동 여부를 체크하는 과정이 번거롭기도 하고,
코드의 변경으로 인해 의도치 않은 사이드 이펙트가 우려되어 코드 수정이 조심스러워 졌던 경험이 있었어요.

그래서.. 이 부분에 있어서 간단한 **단위/통합 테스트**를 작성해보면 어떨까요?
(Vitest로 유틸 함수에 대한 간단한 테스트 코드를 작성했던 것처럼, **Testing Library**를 추가한다면 리액트 컴포넌트와 훅에 대한 테스팅도 가능 하더라구요!)

물론, 모든 곳에 테스트 코드를 작성하는 것은 아니라, **테스트 자동화가 가능하면서 유용하다고 생각되는 부분**만 적용한다면 새로운 기능을 추가하거나 기존의 코드를 수정하는 데 있어서 자신감 있게 개발할 수 있을 것이라고 기대가 돼요!

우선 지금 생각했을 때 적용해보면 괜찮겠다 싶은 내용은 이 정도로 생각했어요:

1. 버튼을 눌렀을 때 의도하는 모달이나 토스트 등이 잘 보여지는가?
2. form 내용을 작성하고 완료 버튼을 눌렀을 때 정상적으로 API 호출하는 동작이 일어나는가?
    - 또한, API 호출의 예외 처리에 페이지를 이동하는 등의 동작이 있다면, 제대로 작동하는가?

테스트 방법에 대해서 몇 가지 찾아봤는데 정리해보니:

- **Vitest + Testing Library로 단위/통합 테스트**
    - 컴포넌트, 훅, 유틸 함수 등의 동작을 테스트 가능.
    - API 호출은 msw로 모킹, zustand 스토어를 사용하는 부분은 [Zustand 공식 문서](https://docs.pmnd.rs/zustand/guides/testing)를 참고해서 모킹 가능.
    - 단점: 로직에 대한 부분은 테스트 자동화는 가능하지만, 화면이 보이는 것은 아니기 때문에 CSS나 레이아웃까지 검증하는 것은 불가능.
        - (이 부분은 Storybook으로 체크하면 좋을 듯 해요!)

- **Storybook으로 UI 테스트 (이미 적용중)**
    - 여러 컴포넌트의 모습을 시각적으로 확인 가능.
    - Chromatic의 시각적 회귀 테스트로 UI의 변경 사항을 체크 가능.
        - 변경된 스토리의 이전 스냅샷과 현재 스냅샷을 리뷰어가 확인하고 approve 하는 기능.
        - 지금은 auto approve 해 놓았는데, 필요에 따라서 PR 리뷰 하면서 변경된 스토리 북 체크하는 과정도 가능하긴 할 것 같아요.

- **Cypress로 E2E 테스트**
    - 사용자가 실제로 브라우저에서 인터랙션을 발생하는 것처럼 BE 서버까지 API 호출을 날려보면서, 어떻게 동작하는지 체크 가능.

    - Cypress 테스트용 서버/DB가 필요해지기도 하고.. BE 팀과의 합을 맞춰야 하는 부분이 많고 자원도 더 필요해질 것이기에 적용하기엔 어렵다는 생각이 들었어요. (아직 e2e 테스트를 적용할 정도로 앱의 규모가 크지도 않기도 하고..?)

요약하자면.. Storybook으로는 UI를 체크하고, Vitest + Testing Library로는 로직을 체크하는 방식 어떨까요??
**즉, Testing Library를 추가로 도입해보면 어떨까요?**

장점으로는 자연스럽게 테스트하기 좋은 코드를 고민하면서 컴포넌트의 역할이 잘 분리되었는지 자체 검증하는 효과도 누릴 수 있을 것으로 예상해요!

**Comment by easyhyun00** [comment]
네! 맞아요 사람이 직접 수동으로 하나씩 체크하는게 번거롭기는 하더라구요.

> Vitest + Testing Library로 단위/통합 테스트
컴포넌트, 훅, 유틸 함수 등의 동작을 테스트 가능.

넵! 저희가 리팩토링이나 새로운 기능을 추가하다보면 기존 로직에서 실수로 놓친 부분이 발생할 수 있을 거 같아요! 그래서 일부 정상적으로 작동을 안 하는 부분이 발생할 수 있을 거 같아 컴포넌트, 훅, 유틸 함수 등 동작을 테스트 하는 게 좋아보입니다!

> Chromatic의 시각적 회귀 테스트로 UI의 변경 사항을 체크 가능.

오! 찾아보니깐 UI 변경 사항을 크로마틱으로 감지할 수 있네요! 그래서 바뀐 UI 가 의도적으로 바뀐건지 실수로 바뀐건지 체크할 수 있어 좋을 것 같습니다!

>Cypress 테스트용 서버/DB가 필요해지기도 하고.. BE 팀과의 합을 맞춰야 하는 부분이 많고 자원도 더 필요해질 것이기에 적용하기엔 어렵다는 생각이 들었어요. (아직 e2e 테스트를 적용할 정도로 앱의 규모가 크지도 않기도 하고..?)

E2E 는 검색해보니 시간이 오래걸리고 자원이 소모된다는 단점이 있네요. 말씀대로 E2E 는 저희가 규모가 큰 편이 아니라 적용을 하지 않아도 좋을 것 같습니다!

> 요약하자면.. Storybook으로는 UI를 체크하고, Vitest + Testing Library로는 로직을 체크하는 방식 어떨까요??
즉, Testing Library를 추가로 도입해보면 어떨까요?

넵 좋은 것 같습니다! 계속 프로젝트를 진행해야 하니 테스팅을 추가하는게 좋을 것 같습니다. 테스트를 작성하는 게 시간은 걸리겠지만 개발 생산성이 향상되고 유지 보수성을 향상 시킬 수 있어 좋은 것 같습니다! Testing Library 에 대해 알아봐야겠네요!!
